### PR TITLE
LAB-220: Unit tests for timeout function

### PR DESCRIPTION
**Pull request by gita135**
Wrote series of unit tests comparing `timeout1()` and `timeout2()`. Both execute successfully  in every case
Compared timings of execution of each to compare accuracy
Accuracy for each function is comparable, either implementation would suffice IMO.

### CODE REVIEW

File: test/runtests.jl [hunk 1]
```diff
@@ -142,6 +142,10 @@ const A = u"A"
     end
 
     @testset "timeout1 timeout2 unit tests" begin
+        function sleep_function(sleep_time::Float64)
```
Comment (mofii) [inline test/runtests.jl:145]: Please squash this commit (d7055d133cda095b236a9acea3ae1c77e3106e1d) with the previous commit (7454d25d5b81d6a87b5742e9159fdf0dc3fad355 ) to make it easier to review the PR.

File: test/runtests.jl [hunk 2]
```diff
@@ -140,4 +140,78 @@ const A = u"A"
             @test after ≈ true_scale
         end
     end
+
+    @testset "timeout1 timeout2 unit tests" begin
```
Comment (mofii) [inline test/runtests.jl:144]: Let's put all the timeout() tests into a separate file and call it `test_timeout.jl` and include it from here.

File: test/runtests.jl [hunk 3]
```diff
@@ -140,4 +140,78 @@ const A = u"A"
...
+        sleep_time = 0.5
+        timeout = 0.3
+        f() = sleep_function(sleep_time)
+        for i in 1:100
+            @test isnothing(TcpInstruments.timeout1(f, timeout)) # Successful timeout/no retval
+            @test isnothing(TcpInstruments.timeout2(f, timeout)) # Successful timeout/no retval
+        end
+        timeout1_time = @elapsed TcpInstruments.timeout1(f, timeout) 
+        timeout2_time = @elapsed TcpInstruments.timeout2(f, timeout) 
+        @info  "timeout1_time < timeout2_time" timeout1_time < timeout2_time
```
Comment (mofii) [inline test/runtests.jl:159]: What do we learn by printing these messages?

File: test/runtests.jl [hunk 4]
```diff
@@ -140,4 +140,78 @@ const A = u"A"
...
+        function sleep_function(sleep_time::Float64)
+            sleep(sleep_time)
+            return 0
+        end
+        # Failed execution --> timeout 
+        sleep_time = 0.5
+        timeout = 0.3
+        f() = sleep_function(sleep_time)
+        for i in 1:100
+            @test isnothing(TcpInstruments.timeout1(f, timeout)) # Successful timeout/no retval
```
Comment (mofii) [inline test/runtests.jl:154]: What's the comment 
```
# Successful timeout/no retval
```
There for? It's always the same comment whether or not the `sleep()` function returned within the timeout tim.

File: test/runtests.jl [hunk 5]
```diff
@@ -140,4 +140,78 @@ const A = u"A"
...
         end
     end
+
+    @testset "timeout1 timeout2 unit tests" begin
+        function sleep_function(sleep_time::Float64)
+            sleep(sleep_time)
+            return 0
+        end
+        # Failed execution --> timeout 
+        sleep_time = 0.5
```
Comment (mofii) [inline test/runtests.jl:150]: Please user shorter sleep times to speed up the tests. Slow tests means that they won't be used be the user, and we should test often when developing code. 
Besides, we have a limited amount of CI minutes per month.
Testing in milliseconds should work just as well as testing in seconds.

**Review by mofii — CHANGES_REQUESTED** [review]
I get thousands of messages like this in the terminal when running these tests:
```julia
.┌ Error: Timed out after 0.1 s.
└ @ TcpInstruments ~/local_src/TcpInstruments.jl/src/util.jl:422
.┌ Error: Timed out after 0.1 s
└ @ TcpInstruments ~/local_src/TcpInstruments.jl/src/util.jl:384
.┌ Error: Timed out after 0.1 s.
└ @ TcpInstruments ~/local_src/TcpInstruments.jl/src/util.jl:422
.┌ Error: Timed out after 0.1 s
└ @ TcpInstruments ~/local_src/TcpInstruments.jl/src/util.jl:384
.┌ Error: Timed out after 0.1 s.
└ @ TcpInstruments ~/local_src/TcpInstruments.jl/src/util.jl:422
.┌ Error: Timed out after 0.1 s
└ @ TcpInstruments ~/local_src/TcpInstruments.jl/src/util.jl:384
.┌ Error: Timed out after 0.1 s.
└ @ TcpInstruments ~/local_src/TcpInstruments.jl/src/util.jl:422
.┌ Error: Timed out after 0.1 s
└ @ TcpInstruments ~/local_src/TcpInstruments.jl/src/util.jl:384
┌ Error: Timed out after 0.1 s.
└ @ TcpInstruments ~/local_src/TcpInstruments.jl/src/util.jl:422

```

Please suppress all of those error messages.

File: Project.toml [hunk 6]
```diff
@@ -4,6 +4,7 @@ authors = ["Morten F. Rasmussen <mf@mofi.dk>, Steve Leung <steve@orchardultrasou
 version = "0.11.13"
 
 [deps]
+BenchmarkTools = "6e4b80f9-dd63-53aa-95a3-0cdb28fa8baf"
```
[the remark this answers is not in the document]
Comment (gita135) [reply Project.toml]: Since both timeout() function time executions are the same, no need to use @elpased now

**Comment by gita135** [comment]
> 

Turns out both `timeout()` functions were using the `@error` which is actually a  Logging macro, rather than throwing an `error` , which is why the errors were logging in the REPL. Now the fix to the proper `error` are being handled appropriately. This will also allow `clear_buffer()` to not throw an error (would need to add try/catch, but I can do that separately PR)

File: test/runtests.jl (hunk 3 above)
Comment (gita135) [reply test/runtests.jl]: Removed them since timing is comparable

**Review by mofii — APPROVED** [review]
👍

**Pull request merged**